Core/MIPS/IR/IRInst.cpp

```diff
@@ -0,0 +1,589 @@
+			break;
+		case IROp::ExitToConstIfGtZ:
+			if ((s32)mips->r[inst->src1] > 0)
+				return constPool[inst->dest];
+			break;
+		case IROp::ExitToConstIfGeZ:
+			if ((s32)mips->r[inst->src1] >= 0)
+				return constPool[inst->dest];
+			break;
+		case IROp::ExitToConstIfLtZ:
+			if ((s32)mips->r[inst->src1] < 0)
+				return constPool[inst->dest];
+			break;
+		case IROp::ExitToConstIfLeZ:
+			if ((s32)mips->r[inst->src1] <= 0)
+				return constPool[inst->dest];
+			break;
+
+		case IROp::Downcount:
+			mips->downcount -= (inst->src1) | ((inst->src2) << 8);
+			break;
+
+		case IROp::SetPC:
+			mips->pc = mips->r[inst->src1];
+			break;
+
+		case IROp::SetPCConst:
+			mips->pc = constPool[inst->src1];
+			break;
+
+		case IROp::Syscall:
+			// SetPC was executed before.
+		{
+			MIPSOpcode op(constPool[inst->src1]);
+			CallSyscall(op);
+			return mips->pc;
+		}
+
+		case IROp::Interpret:  // SLOW fallback. Can be made faster.
+		{
+			MIPSOpcode op(constPool[inst->src1]);
+			MIPSInterpret(op);
+			break;
+		}
+
+		case IROp::CallReplacement:
+		{
+			int funcIndex = constPool[inst->src1];
+			const ReplacementTableEntry *f = GetReplacementFunc(funcIndex);
+			int cycles = f->replaceFunc();
+			mips->downcount -= cycles;
+			return mips->r[MIPS_REG_RA];
+		}
+
+		case IROp::Break:
+			Crash();
+			break;
+
+		case IROp::SetCtrlVFPU:
+			mips->vfpuCtrl[inst->dest] = constPool[inst->src1];
+			break;
+
+		default:
+			Crash();
+		}
+#ifdef _DEBUG
+		if (mips->r[0] != 0)
+			Crash();
+#endif
+		inst++;
+	}
+
+	// If we got here, the block was badly constructed.
+	Crash();
+	return 0;
+}
+
+void IRWriter::Write(IROp op, u8 dst, u8 src1, u8 src2) {
+	IRInst inst;
+	inst.op = op;
+	inst.dest = dst;
+	inst.src1 = src1;
+	inst.src2 = src2;
+	insts_.push_back(inst);
+}
+
+void IRWriter::WriteSetConstant(u8 dst, u32 value) {
+	Write(IROp::SetConst, dst, AddConstant(value));
+}
+
+int IRWriter::AddConstant(u32 value) {
+	for (size_t i = 0; i < constPool_.size(); i++) {
+		if (constPool_[i] == value)
+			return (int)i;
+	}
+	constPool_.push_back(value);
+	if (constPool_.size() > 255) {
+		// Cannot have more than 256 constants in a block!
+		Crash();
+	}
+	return (int)constPool_.size() - 1;
+}
+
+int IRWriter::AddConstantFloat(float value) {
+	u32 val;
+	memcpy(&val, &value, 4);
+	return AddConstant(val);
+}
+
+void IRWriter::Simplify() {
+	SimplifyInPlace(&insts_[0], (int)insts_.size(), constPool_.data());
+}
+
+const char *GetGPRName(int r) {
+	if (r < 32) {
+		return currentDebugMIPS->GetRegName(0, r);
+	}
+	switch (r) {
+	case IRTEMP_0: return "irtemp0";
+	case IRTEMP_1: return "irtemp1";
+	case IRTEMP_LHS: return "irtemp_lhs";
+	case IRTEMP_RHS: return "irtemp_rhs";
+	default: return "(unk)";
+	}
+}
+
+void DisassembleParam(char *buf, int bufSize, u8 param, char type, const u32 *constPool) {
+	switch (type) {
+	case 'G':
+		snprintf(buf, bufSize, "%s", GetGPRName(param));
+		break;
+	case 'F':
+		snprintf(buf, bufSize, "r%d", param);
```

Comment:
Hmm, not `f`?  Or I guess even `currentDebugMIPS->GetRegName(1, param)`?

-[Unknown]